Sources/Event.swift

```diff
@@ -861,6 +861,48 @@ extension Signal.Event {
 			}
 		}
 	}
+	
+	internal static func collect(every interval: DispatchTimeInterval, on scheduler: DateScheduler, skipEmpty: Bool, discardsWhenCompleted: Bool) -> Transformation<[Value], Error> {
+		return { action, lifetime in
+			let values = Atomic<[Value]>([])
+			let d = SerialDisposable()
+			var lastDelivered: Date = scheduler.currentDate
+			
+			d.inner = scheduler.schedule(after: scheduler.currentDate.addingTimeInterval(interval), interval: interval, leeway: interval * 0.1) {
+				let currentValues = values.swap([])
+				if !(currentValues.isEmpty && skipEmpty) {
+					action(.value(currentValues))
+					lastDelivered = scheduler.currentDate
+				}
+			}
+			
+			lifetime.observeEnded {
+				d.dispose()
+				scheduler.schedule { action(.interrupted) }
+			}
+
+			return { event in
+				switch event {
+				case let .value(value):
+					values.modify { $0.append(value) }
+				case let .failed(error):
+					d.inner = scheduler.schedule { action(.failed(error)) }
+				case .completed:
+					let currentValues = values.swap([])
+					if !discardsWhenCompleted, !(currentValues.isEmpty && skipEmpty) {
+						d.inner = scheduler.schedule(after: lastDelivered.addingTimeInterval(interval)) {
```

Comment:
This `schedule` call makes me a little uneasy. It's difficult to reason about how this will interact with the other scheduling call. (e.g. it's possible that the above schedule could run in-between this `values.swap` and this `scheduler.schedule`).

I think we should solve this by removing this `schedule` and sending the complete through the top `schedule`. This would be best done by adding a `CollectEveryState` struct and putting that inside an `Atomic` instead of a `Atomic<[Value]>`.

Something like this:

```swift
state CollectEveryState<Value> {
    var skipEmpty: Bool
    var values: [Value] = []
    var isCompleted: Bool

    var hasValues: Bool {
       return !values.isEmpty || !skipEmpty
    }

    func collect() -> [Value]? {
        if !hasValues {
            return nil
        }
        defer { values.removeAll() }
        return values
    }
}
```

That should clean up the implementation a bit too, which is nice.

Reply:
@mdiep Done 👍